standard/expressions.md

```diff
@@ -2221,29 +2218,94 @@ The *primary_expression* of an *element_access* shall not be an *array_creation_
 
 #### 12.8.12.2 Array access
 
-For an array access, the *primary_expression* of the *element_access* shall be a value of an *array_type*. Furthermore, the *argument_list* of an array access shall not contain named arguments. The number of expressions in the *argument_list* shall be the same as the rank of the *array_type*, and each expression shall be of type `int`, `uint`, `long`, or `ulong,` or shall be implicitly convertible to one or more of these types.
+For an array access the *argument_list* shall not contain named arguments or by-reference arguments (§15.6.2.3).
+
+The number of expressions in the *argument_list* shall be the same as the rank of the *array_type*, and each expression shall be:
+
+- of type `int`, `uint`, `long`, or `ulong`; or
+- for single rank array access only, of type `Index` or `Range`; or
+- be implicitly convertible to one or more of the above types.
 
-The result of evaluating an array access is a variable of the element type of the array, namely the array element selected by the values of the expressions in the *argument_list*.
+The run-time processing of an array access of the form `P[A]`, where `P` is a *primary_expression* of an *array_type* and `A` is an *argument_list* of index expressions, consists of the following steps:
 
-The run-time processing of an array access of the form `P[A]`, where `P` is a *primary_expression* of an *array_type* and `A` is an *argument_list*, consists of the following steps:
+- `P` is evaluated. If this evaluation causes an exception, no further steps are executed.
+- For each index expression in the *argument_list* in order, from left to right:
+  - The index expression is evaluated, let the type of the resultant value be *T*;
+  - This value is then converted to the first of the types: `int`, `uint`, `long`, `ulong`, or for single rank array access only, `Index` or `Range`; for which an implicit conversion ([§10.2](conversions.md#102-implicit-conversions)) from *T* exists.
+  - If evaluation of an index expression or the subsequent implicit conversion causes an exception, then no further index expressions are evaluated and no further steps are executed.
+- The value of `P` is checked to be valid. If the value of `P` is `null`, a `System.NullReferenceException` is thrown and no further steps are executed.
+- If the preceding steps have produced a single index value of type `Range` then:
+  - Let *L* be the length of the array referenced by `P`.
+  - `A` is checked to be valid with respect to *L* (§24.3), if it is not then a `System.ArgumentOutOfRangeException` is thrown and no further steps are executed.
+  - The starting offset, *S*, and number of items, *N*, for `A` with respect to *L* are determined as described for `GetOffsetAndLength` (§24.3).
+  - A new array is created from a shallow copy of the *N* elements of `P` starting at index *S*, if *N* is zero the new array has zero elements. This array becomes the result of the array access.
```

Comment:
It's not required that a new array is created in the case where the range has zero elements. The current compiler defers this choice to RuntimeHelpers.GetSubArray, which returns a cached empty array instance if the range turns out to have zero length at runtime. (https://github.com/dotnet/runtime/blob/v9.0.8/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs#L26-L44)

Reply:
We want to allow (but not require) the special-casing for a zero-size array, but a requirement for non-zero-size arrays.

We're okay with either this being fixed in this PR in a new commit, or @Nigel-Ecma raising a new issue and explicitly deferring.

Reply:
The Standard does not currently specify this behavior in other situations, e.g. for array creation (§12.8.16.5), so the issue needs to be addressed globally and not just here. PR #1403 has been spun off and this thread will be marked as resolved.